### PR TITLE
Cloud Code Modernization

### LINKED ISSUE #4930: Modernization of hooks and functions
**Is your feature request related to a problem? Please describe.**

Right now, we still rely on the legacy backbone style success / error callback methods. With node.js moving forward at an incredible pace, I believe it's time to make the cloud functions enter the 21st century supporting async / await, promises and more

**Describe the solution you'd like**

- Support promises, async/await in cloud code as valid return values for hooks and functions
- Mark calls to success / error callbacks deprecated in Cloud Code (with a big fat warning)
- Later on, remove success / error implementations

Cloud funcitons would look like:

```
Parse.Cloud.beforeSave('ClassName', async ({ object }) => {
    await doSomething();
    if (somethingsFishy(object)) {
        throw new Error('OOps')
    }
    return;
});

// Also valid
Parse.Cloud.beforeSave('ClassName', async ({ object }) => {
    await doSomething();
    if (somethingsFishy(object)) {
        return new Error('OOps'); // returning an error is also allowed
    }
});
```

For functions:

```
Parse.Cloud.define('MyFunction', async ({ params }) => {
    await doSomething();
    if (!param.isValid) {
        return new Error('OOps'); // returning an error is also allowed
    }
    return 'OK!';
});
```

cc: @acinader , @dplewis

### PR DESCRIPTION
**Pull request by flovilmart**
Closes #4930 

- Removes success / error based callbacks
- Removes `response` object from cloud code hooks and functions
- Removes `status` object from cloud code jobs
- Support fully promise and async await based APIs

### CODE REVIEW

File: 3.0.0.md [hunk 1]
```diff
@@ -0,0 +1,209 @@
+# Upgrading Parse Server to version 3.0.0
+
+With the 3.0.0, parse-server has completely revamped it's cloud code interface. Gone are the backbone style calls, welcome promises and async await.
+
+In order to leverage those new nodejs features, you'll need to run at least node 8.10. We've dropped the support of node 6 a few months ago, so you should feed really comfortable at home.
```
Comment (acinader) [inline 3.0.0.md]: "so you should feed really comfortable at home." is a typo.  maybe:  "so if you haven't made the change yet, now would be a really good time to take the plunge."

File: 3.0.0.md [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+// after
+Parse.Cloud.beforeSave('MyClassName', (request) => {
+  if (!passesValidation(req.object)) {
+    throw 'Ooops something went wrong'; 
+  }
+});
+```
+
+All methods are wrapped into promises, so you can freely `throw` `Error`, `Parse.Error` or `string` to mark an error.
```
Comment (acinader) [inline 3.0.0.md]: "wrapped into" should be "wrapped in".

File: 3.0.0.md [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    url: request.object.get('fileURL')
+  });
+
+  const file = new Parse.File('image.png', { base64: contents.buffer.toString('base64') });
+  request.object.set('file', file);
+  await file.save();
+});
+```
+
+Note: whenever a failure will occur, the beforeSave will fail.
```
Comment (acinader) [inline 3.0.0.md]: Note: if an error occurs or a string is thrown, the beforeSave will fail to save the state of the object being saved.

File: 3.0.0.md [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+## Migrating Functions
+
+Cloud Functions can be migrated the same way as cloud code hooks.
+In functions, the response object is not passed anymore, as it is in cloud code hooks
+
+Let's get again with the image downloader:
```
Comment (acinader) [inline 3.0.0.md]: Continuing with the image downloader exampe:

File: 3.0.0.md [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+
+  const file = new Parse.File(name, { base64: response.buffer.toString('base64') });
+  await file.save();
+  return file;
+});
+```
+
+## Migrating jobs
+
+As the hooks and functions, jobs don't have a status object anymore.
```
Comment (acinader) [inline 3.0.0.md]: As with hooks and functions....

File: 3.0.0.md [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+The message method was moved to the request object.
```
Comment (acinader) [inline 3.0.0.md]: The message method has been moved to the request object.

File: 3.0.0.md [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+    const file = new Parse.File(image.get('name'), { base64: contents.buffer.toString('base64') });
+    await file.save();
+    request.message('Saved ' + image.get('url'));
+  });
+  await Promise.all(promises);
+});
+
+```
+
+As you can see the new implementation is way more concise and easier to read and maintain.
```
Comment (acinader) [inline 3.0.0.md]: As you can see, the new implementation is more concise, easier to read and maintain.

File: 3.0.0.md [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+
+As always if you encounter a problem, discover an issue with this guide or somewhere else, feel free to [reach out on github](https://github.com/parse-community/parse-server/issues/new/choose)
```
Comment (acinader) [inline 3.0.0.md]: If you encounter a problem or discover an issue, with this guide or with any Parse Community product,  feel free to [reach out on github](https://github.com/parse-community/parse-server/issues/new/choose)

**Comment by flovilmart** [comment]
thanks @acinader, I went through it. Overall is the feature as expected?

**Comment by acinader** [comment]
yeah, it looks great.  will need to revise the guides.

I'd like to use this branch and conform my cloud code to get a better idea of how it works, but I don't think I'll be able to do that until next week.  I'll try to do sooner.

File: README.md [hunk 9]
```diff
@@ -380,6 +381,14 @@ Parse Server allows developers to choose from several options when hosting files
 
 `GridStoreAdapter` is used by default and requires no setup, but if you're interested in using S3 or Google Cloud Storage, additional configuration information is available in the [Parse Server guide](http://docs.parseplatform.org/parse-server/guide/#configuring-file-adapters).
 
+# Upgrading to 3.0.0
+
+Starting 3.0.0, parse-server uses the JS SDK version 2.0. 
+In short, parse SDK v2.0 removes the backbone style callbacks as well as the Parse.Promise object in favor of native promises.
+All the Cloud Code interfaces also have been updated to reflect those changes, and all backbone style response objects are removed and replaced by Promise style resolution.
+
+We written up a [migration guide](3.0.0.md), hoping this will help you transition to the next major release.
```
Comment (acinader) [inline README.md]: We have written up.....

**Comment by flovilmart** [comment]
@acinader I added the small nit, if you can add back the seal of approval :)

**Comment by flovilmart** [comment]
@acinader do you think we should merge with master?

**Comment by dplewis** [comment]
@flovilmart I’d also like to use this branch and use it in production first

**Comment by flovilmart** [comment]
@dplewis yep no problem! Let me know if there's any issue.

**Comment by dplewis** [comment]
I’ll have time tomorrow. Blog post?

**Comment by flovilmart** [comment]
Would be good! JS SDK v2.0 and server v3.0, may not have time for a blog post write up any one of you would?

**Comment by flovilmart** [comment]
@dplewis should we merge this one on master soon?

**Comment by flovilmart** [comment]
I've rebased it all so I'll be able to rebase and merge cleanly

**Comment by dplewis** [comment]
I’m about half way through rewriting my cloud code. No issues so far.

**Comment by flovilmart** [comment]
@acinader @dplewis are we good with this one?

**Comment by acinader** [comment]
I haven't done the upgrade yet, but yes, I am good with it.  let's get it out there.

**Pull request merged**